server/encryption/localkey_handler.go

```diff
@@ -0,0 +1,217 @@
+	wrappededKey, err := handler.keywrapper.Wrap(dks)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return wrappededKey, nil
+
+}
+
+func (handler *LocalEncryptionHandler) unwrapDKS(wrappedDKS []byte) ([]byte, error) {
+	// use Tinker to unwrap Data Key
+	// https://github.com/google/tink/commit/22467ef7273d73b2d65e4b50310aab4af006bb7e
+	key, err := handler.keywrapper.Unwrap(wrappedDKS)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return key, nil
+}
+
+func (handler *LocalEncryptionHandler) encryptData(dks []byte, plaintextData []byte) ([]byte, error) {
+	// Init cypher in GCM
+
+	block, err := aes.NewCipher(dks)
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+
+	if err != nil {
+		return nil, err
+	}
+	// init nonce
+	nonce := make([]byte, gcm.NonceSize())
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		panic(err.Error())
+	}
+	ciphertext := gcm.Seal(nonce, nonce, plaintextData, nil)
+	return ciphertext, nil
+}
+
+func (handler *LocalEncryptionHandler) decryptData(dks []byte, encryptedData []byte) ([]byte, error) {
+
+	block, err := aes.NewCipher(dks)
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// get nonce
+	nonceSize := gcm.NonceSize()
+	nonce, ciphertext := encryptedData[:nonceSize], encryptedData[nonceSize:]
+
+	//Decrypt the data
+
+	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return plaintext, nil
+}
+
+// Seal takes the message, performs the encryption and return
+// the encrypted data along with the wrapped data
+// The encoded message contains the first byte as size of the wrapped key,
+// the wrapped key and finally the encrypted message.
+
+// |  byte 0  |   byte 1   |   byte 2   |    ...   | byte (n +1 ) |    byte (n+2)  |  ... | byte (n + m + 2) |
+// |----------|------------|------------|----------|--------------|----------------|------|------------------|
+// | key size | key byte 0 | key byte 1 |      ... | key byte n   | message byte 0 |  ... |  message byte m  |
+
+func (handler *LocalEncryptionHandler) Seal(data []byte) ([]byte, error) {
+	// Generate a default Data Key (DKS) if not yet available
+	if handler.defaultDKS == nil {
+		dksKey, err := handler.generateDKS()
+
+		if err != nil {
+			return nil, err
+		}
+		handler.defaultDKS = dksKey
+	}
+
+	// Cipher the message
+	ciphertext, err := handler.encryptData(handler.defaultDKS, data)
+	if err != nil {
+		return nil, err
+	}
+
+	// wrap the data key
+
+	wrappedKey, err := handler.wrapDKS(handler.defaultDKS)
+	if err != nil {
+		return nil, err
+	}
+
+	// concate wrapped key with cipyer text
+
+	keyLength := len(wrappedKey)
+
+	// concate key size | wrapped key | ciphertext
+	keySize := []byte{byte(keyLength)}
+	dataSequence := append(keySize, wrappedKey...)
+	dataSequence = append(dataSequence, ciphertext...)
```

Comment:
Would be more efficient to pre-allocate the slice and copy the data in I think.

Reply:
agreed. wip